hypershift-operator/controllers/nodepool/kubevirt/imagecacher.go

```diff
@@ -0,0 +1,263 @@
+// reference to be used by the node templates.
+type BootImage interface {
+	// CacheImage creates a PVC to cache the node image.
+	CacheImage(ctx context.Context, cl client.Client, nodePool *hyperv1.NodePool) error
+	getDVSourceForVMTemplate() *v1beta1.DataVolumeSource
+}
+
+// containerImage is the implementation of the BootImage interface for container images
+type containerImage struct {
+	name string
+}
+
+func newContainerBootImage(imageName string) *containerImage {
+	return &containerImage{
+		name: imageName,
+	}
+}
+
+func (containerImage) CacheImage(_ context.Context, _ client.Client, _ *hyperv1.NodePool) error {
+	return nil // no implementation
+}
+
+func (ci containerImage) getDVSourceForVMTemplate() *v1beta1.DataVolumeSource {
+	pullMethod := v1beta1.RegistryPullNode
+	return &v1beta1.DataVolumeSource{
+		Registry: &v1beta1.DataVolumeSourceRegistry{
+			URL:        &ci.name,
+			PullMethod: &pullMethod,
+		},
+	}
+}
+
+// qcowImage is the implementation of the BootImage interface for QCOW images
+type qcowImage struct {
+	name      string
+	hash      string
+	namespace string
+	dvName    string
+	log       logr.Logger
+}
+
+func newQCOWBootImage(name, hash, namespace string, logger logr.Logger) *qcowImage {
+	return &qcowImage{
+		name:      name,
+		hash:      hash,
+		namespace: namespace,
+		log:       logger,
+		dvName:    getDvName(hash),
+	}
+}
+
+func (qi qcowImage) CacheImage(ctx context.Context, cl client.Client, nodePool *hyperv1.NodePool) error {
+	kvPlatform := nodePool.Spec.Platform.Kubevirt
+
+	objectKey := client.ObjectKey{Namespace: qi.namespace, Name: qi.dvName}
+
+	dvFound, err := dvExists(ctx, cl, objectKey)
+	if err != nil {
+		return fmt.Errorf("failed to read the DataVolume; namespace: %s; name: %s; %w", qi.namespace, qi.dvName, err)
+	}
+
+	pvc := &corev1.PersistentVolumeClaim{}
+	err = cl.Get(ctx, objectKey, pvc)
+	if errors.IsNotFound(err) { // the PVC wasn't found
+		if !dvFound {
+			qi.log.Info("couldn't find boot image cache DataVolume; creating it...", "DataVolume Name", qi.dvName)
+			return qi.createDVForCache(ctx, cl, kvPlatform)
+		} else {
+			qi.log.Info("found the boot image cache DataVolume. The PVC should be created shortly")
+		}
+		return nil // the DV was already created, so we just need to wait for the PVC to be created.
+	}
+
+	if err != nil {
+		return fmt.Errorf("couldn't get boot image cache PVC; %w", err)
+	}
+
+	// PVC exists; label it if not already labeled; This is done because DataVolume labels are not copied to the PVC
+	// See here: https://github.com/kubevirt/containerized-data-importer/issues/2533
+	if _, labelExist := pvc.Labels[bootImagePVCLabelRoleName]; !labelExist {
+		if pvc.Labels == nil {
+			pvc.Labels = make(map[string]string)
+		}
+
+		patch := []byte(fmt.Sprintf(`[{"op":"replace","path":"/metadata/labels/%s","value":"%s"}]`,
+			bootImagePVCLabelRoleNameForPatch,
+			bootImagePVCLabelRoleValue,
```

Comment:
would it make sense to just replace the entire `/metadata/labels` path here with a `test` operation to ensure no collisions occurred?  rather than using the bootImagePVCLabelRoleNameForPatch?

Reply:
Originally, there was anoyher label for the hash. But lsbel value must not be longer than 64 chars. The hash is 64...

Anyway, this is s leftover.